File: script/geo.py
'''
>>>
>>> x, y, R, A, B, C = symbols('x y R A B C')

line = A*x + B*y + C

circle = x**2 + y**2 - R**2

intersection = nonlinsolve([circle, line], [x, y])

x0, y0 = list(intersection)[0]

>>> (x0*A + C)/B
       ____________________
      /  2  2    2  2    2
  A*\/  A *R  + B *R  - C       B*C
- ------------------------- + -------
            2    2             2    2
           A  + B             A  + B
>>> y0
     ____________________
    /  2  2    2  2    2
A*\/  A *R  + B *R  - C       B*C
------------------------- - -------
          2    2             2    2
         A  + B             A  + B
>>>


x1, y1 = list(intersection)[1]

>>> (x1*A + C)/B
     ____________________
    /  2  2    2  2    2
A*\/  A *R  + B *R  - C       B*C
------------------------- + -------
          2    2             2    2
         A  + B             A  + B
>>> y1
       ____________________
      /  2  2    2  2    2
  A*\/  A *R  + B *R  - C       B*C
- ------------------------- - -------
            2    2             2    2
           A  + B             A  + B


ellipse = (x/Rx)**2 + (y/Ry)**2 - 1

eint = nonlinsolve([line, ellipse], [x, y])


>>> (x0*A + C)/B
             ______________________
            /  2   2    2   2    2              2
  A*Rx*Ry*\/  A *Rx  + B *Ry  - C         B*C*Ry
- --------------------------------- + ---------------
            2   2    2   2             2   2    2   2
           A *Rx  + B *Ry             A *Rx  + B *Ry

>>> y0
           ______________________
          /  2   2    2   2    2              2
A*Rx*Ry*\/  A *Rx  + B *Ry  - C         B*C*Ry
--------------------------------- - ---------------
          2   2    2   2             2   2    2   2
         A *Rx  + B *Ry             A *Rx  + B *Ry


>>> (x1*A + C)/B
           ______________________
          /  2   2    2   2    2              2
A*Rx*Ry*\/  A *Rx  + B *Ry  - C         B*C*Ry
--------------------------------- + ---------------
          2   2    2   2             2   2    2   2
         A *Rx  + B *Ry             A *Rx  + B *Ry

>>> y1
             ______________________
            /  2   2    2   2    2              2
  A*Rx*Ry*\/  A *Rx  + B *Ry  - C         B*C*Ry
- --------------------------------- - ---------------
            2   2    2   2             2   2    2   2
           A *Rx  + B *Ry             A *Rx  + B *Ry

'''
from math import sqrt


class Point(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return 'Point({}, {})'.format(self.x, self.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __ne__(self, other):
        return self.x != other.x or self.y != other.y

    def __add__(self, offset):
        return Point(self.x + offset[0], self.y + offset[1])

    def __sub__(self, offset):
        return Point(self.x - offset[0], self.y - offset[1])

    def offset(self, other):
        return (self.x - other.x, self.y - other.y)


class ProjectiveLine(object):
    def __init__(self, A, B, C):
        # Coordinates are in the projective plane
        # such that Ax + By + C = 0
        if (A == 0 and B == 0):
            raise ValueError('Invalid projective line coordinates')
        self.A = A
        self.B = B
        self.C = C
        self.norm2 = self.A**2 + self.B**2

    def __str__(self):
        return 'Line ({}, {}, {})'.format(A, B, C)

    def parallel_line(self, offset):
        return ProjectiveLine(self.A, self.B, self.C + offset*sqrt(self.norm2))

    def distance_from(self, point):
        return abs(point.x*self.A + point.y*self.B + self.C)/sqrt(self.norm2)

    def translate(self, offset):
        return ProjectiveLine(self.A, self.B, self.C - (offset[0]*self.A + offset[1]*self.B))


class LineSegment(ProjectiveLine):
    def __init__(self, start, end):
        if isinstance(start, tuple):
            start = Point(*start)
        if isinstance(end, tuple):
            end = Point(*end)
        super().__init__(end.y - start.y,
                         start.x - end.x,
                         end.x*start.y - start.x*end.y)
        self.start = start
        self.end = end


class Ellipse(object):
    def __init__(self, centre, x_radius, y_radius):
        if x_radius == 0 or y_radius == 0:
            raise ValueError('Invalid radius')
        if isinstance(centre, tuple):
            centre = Point(*centre)
        self.centre = centre
        self.x_radius = x_radius
        self.y_radius = y_radius

    def line_intersect(self, line):
        # Note: formulae were derived by using Sympy to find
        #       intersection points

        # Translate our line so that it corresponds to the ellipse
        # having its centre at the origin
        if self.centre.x != 0 or self.centre.y != 0:
            line = line.translate((-self.centre.x, -self.centre.y))

        if line.A == 0:
            y = -line.C/line.B
            x2 = self.x_radius**2 - (y*self.x_radius/self.y_radius)**2
            y += self.centre.y
            if x2 < 0:
                return []
            elif x2 == 0:
                return [Point(self.centre.x, y)]
            else:
                x = sqrt(x2) + self.centre.x
                return [Point(-x, y), Point(x, y)]
        else:
            d2 = (line.A*self.x_radius)**2 + (line.B*self.y_radius)**2
            a = -(line.C*line.B*self.y_radius**2)/d2
            b = d2 - line.C**2
            if b < 0:
                return []
            elif b == 0:
                return [Point(-(line.C + a*line.B)/line.A, a) + (self.centre.x, self.centre.y)]
            else:
                c = -line.A*self.x_radius*self.y_radius*sqrt(b)/d2
                return [Point(-(line.C + (a + c)*line.B)/line.A, (a + c)) + (self.centre.x, self.centre.y),
                        Point(-(line.C + (a - c)*line.B)/line.A, (a - c)) + (self.centre.x, self.centre.y)]

    def contains(self, point):
        return (((point.x - self.centre.x)/self.x_radius)**2
              + ((point.y - self.centre.y)/self.y_radius)**2) < 1.0

    def translate(self, offset):
        return Ellipse(self.centre + offset, self.x_radius, self.y_radius)


class Circle(Ellipse):
    def __init__(self, centre, radius):
        if isinstance(centre, tuple):
            centre = Point(*centre)
        super().__init__(centre, radius, radius)
        self.radius = radius


class LineString(object):
    def __init__(self, end_points, close=False):
        self._segments = []
        for n in range(len(end_points)-1):
            self._segments.append(LineSegment(end_points[n], end_points[n+1]))
        if close:
            self._segments.append(LineSegment(end_points[-1], end_points[0]))


class Polygon(object):
    def __init__(self, points):
        self._boundary = LineString(points, True)


class Rectangle(Polygon):
    def __init__(self, top_left, bottom_right):
        if isinstance(top_left, tuple):
            top_left = Point(*top_left)
        if isinstance(bottom_right, tuple):
            bottom_right = Point(*bottom_right)
        if top_left == bottom_right:
            raise ValueError('Rectangle has no size')
        self.width = abs(top_left.x - bottom_right.x)
        self.height = abs(top_left.y - bottom_right.y)
        self.centre = Point((top_left.x + bottom_right.x)/2.0,
                            (top_left.y + bottom_right.y)/2.0)
        self.top_left = self.centre - (self.width/2.0, self.height/2.0)
        self.bottom_right = self.centre + (self.width/2.0, self.height/2.0)

    def contains(self, point):
        return ((0 < (point.x - self.centre.x) < self.width)
            and (0 < (point.y - self.centre.y) < self.height))

class RoundedRectangle(Rectangle):
    def __init__(self, top_left, bottom_right, x_corner_radius=0, y_corner_radius=0):
        super().__init__(top_left, bottom_right)
        if y_corner_radius == 0:
            y_corner_radius = x_corner_radius
        if (x_corner_radius < 0 or x_corner_radius > self.width/2.0
         or y_corner_radius < 0 or y_corner_radius > self.height/2.0):
            raise ValueError('Invalid corner radius')
        self.x_corner_radius = x_corner_radius
        self.y_corner_radius = y_corner_radius
        if x_corner_radius == 0 and y_corner_radius == 0:
            self.inner_rectangle = Rectangle(top_left, bottom_right)
            self.corner_ellipse = None
        else:
            w_2 = (self.width - x_corner_radius)/2.0
            h_2 = (self.height - y_corner_radius)/2.0
            self.inner_rectangle = Rectangle(Point(self.centre.x) - (w_2, h_2),
                                             Point(self.centre.x) + (w_2, h_2))
            self.corner_ellipse = Ellipse(self.centre, x_corner_radius, y_corner_radius)

    def contains(self, point):
        if x_corner_radius == 0 and y_corner_radius == 0:
            return super().contains(point)
        else:
            return (self.inner_rectangle.contains(point)
                 or self.corner_ellipse.translate(-self.width/2.0, -self.height/2.0).contains(point)
                 or self.corner_ellipse.translate(-self.width/2.0,  self.height/2.0).contains(point)
                 or self.corner_ellipse.translate( self.width/2.0, -self.height/2.0).contains(point)
                 or self.corner_ellipse.translate( self.width/2.0,  self.height/2.0).contains(point))


if __name__ == '__main__':
    c0 = Circle(Point(0, 0), 8)
    cx = Circle(Point(2, 0), 8)
    cy = Circle(Point(0, 2), 8)
    circles = [c0, cx, cy]

    l0 = LineSegment(Point(0, -1), (0, 1))
    lx = LineSegment(Point(2, -1), (2, 1))
    ly = LineSegment(Point(-1, 2), (1, 2))
    lines = [l0, lx, ly]

    for l in lines:
        for c in circles:
            print([str(p) for p in c.line_intersect(l)])

File: script/geo.js
/******************************************************************************

Cell Diagramming Language

Copyright (c) 2018  David Brooks

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

******************************************************************************/

'use strict';

//==============================================================================

import * as exception from './exception.js';
import {setAttributes, List} from './utils.js';

// TEMP
const SVG_NS = 'http://www.w3.org/2000/svg';

//==============================================================================

// Two numbers are considered equal if closer than EPISILON

const EPISILON = 1.0e-6

//==============================================================================

class GeoObject
{
    drawSvg(parentNode)
    {
    }
}

//==============================================================================

export class Point extends GeoObject
{
    constructor(x, y)
    {
        super();
        this.x = x;
        this.y = y;
    }

    toString()
    {
        return `Point(${this.x}, ${this.y})`;
    }

    equal(other)
    {
        return (Math.abs(this.x - other.x) < EPISILON)
            && (Math.abs(this.y - other.y) < EPISILON);
    }

    notEqual(other)
    {
        return !this.equal(other)
    }

    add(offset)
    {
        return new Point((this.x + offset[0]), (this.y + offset[1]));
    }

    subtract(offset)
    {
        return new Point((this.x - offset[0]), (this.y - offset[1]));
    }

    offset(other)
    {
        return [(this.x - other.x), (this.y - other.y)];
    }

    distance(other)
    {
        return Math.sqrt(Math.pow((this.x - other.x), 2)
                       + Math.pow((this.y - other.y), 2));
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'circle');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 cx: this.x, cy: this.y, r: 4,
                                 fill: 'red'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class ProjectiveLine extends GeoObject
{
    constructor(A, B, C)
    {
        if ((A === 0) && (B === 0)) {
            throw new exception.ValueError("Invalid projective line coordinates");
        }
        super();
        this.A = A;
        this.B = B;
        this.C = C;
        this.normSquared = Math.pow(this.A, 2) + Math.pow(this.B, 2);
    }

    toString() {
        return `Line (${A}, ${B}, ${C})`;
    }

    contains(point)
    {
        return Math.abs(this.A*point.x + this.B*point.y + this.C) < EPISILON;
    }

    parallelLine(offset) {
        return new ProjectiveLine(this.A, this.B, this.C + offset*Math.sqrt(this.normSquared));
    }

    distanceFrom(point) {
        return Math.abs(point.x*this.A + point.y*this.B + this.C)/Math.sqrt(this.normSquared);
    }

    translate(offset) {
        return new ProjectiveLine(this.A, this.B, this.C - (offset[0]*this.A + offset[1]*this.B));
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'path');
        // check that A and B are both non-zero
        const start = new Point(0, -this.C/this.B);
        const end = new Point(-this.C/this.A, 0);
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 d: `M${start.x},${start.y}L${end.x},${end.y}`,
                                 stroke: 'green'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class LineSegment extends ProjectiveLine
{
    constructor(start, end)
    {
        if (start instanceof Array) {
            start = new Point(...start);
        }
        if (end instanceof Array) {
            end = new Point(...end);
        }
        super(end.y - start.y, start.x - end.x, end.x*start.y - start.x*end.y);
        this.length = Math.sqrt(this.normSquared);
        this.start = start;
        this.end = end;
    }

    contains(point)
    {
        return super.contains(point)
            && (this.start.distance(point) <= this.length)
            && (this.start.distance(point) <= this.length);
    }

    translate(offset) {
        return new LineSegment(this.start.add(offset), this.end.add(offset));
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'path');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 d: `M${this.start.x},${this.start.y}L${this.end.x},${this.end.y}`,
                                 stroke: 'green'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class LineString extends GeoObject
{
    constructor(endPoints, closed=false) {
        super();
        this._segments = [];
        this.coordinates = []
        if (endPoints.length > 0) {
            for (let n = 0; n < (endPoints.length - 1); n += 1) {
                const segment = new LineSegment(endPoints[n], endPoints[n + 1]);
                this._segments.push(segment);
                this.coordinates.push(segment.start);
            }
            if (closed) {
                const segment = new LineSegment(endPoints.slice(-1)[0], endPoints[0]);
                this._segments.push(segment);
                this.coordinates.push(segment.start);
            }
            this.coordinates.push(this._segments.slice(-1)[0].end);
        }
    }

    drawSvg(parentNode)
    {
        const points = this.coordinates;
        let pointCoords = [];
        for (let point of points.slice(1)) {
            pointCoords.push(`L${point.x},${point.y}`);
        }
        const svgNode = document.createElementNS(SVG_NS, 'path');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 d: `M${points[0].x},${points[0].y}${pointCoords.join('')}`});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class Polygon extends GeoObject
{
    constructor(points)
    {
        super();
        this._boundary = new LineString(points, true);
    }

    drawSvg(parentNode)
    {
        const points = this._boundary.coordinates;
        let pointCoords = [];
        for (let point of points.slice(1)) {
            pointCoords.push(`${point.x},${point.y}`);
        }
        const svgNode = document.createElementNS(SVG_NS, 'polygon');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 points: `${points[0].x},${points[0].y} ${pointCoords.join(' ')}`,
                                 fill: 'none'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class Rectangle extends GeoObject
{
    constructor(topLeft, bottomRight)
    {
        if (topLeft instanceof Array) {
            topLeft = new Point(...topLeft);
        }
        if (bottomRight instanceof Array) {
            bottomRight = new Point(...bottomRight);
        }
        if ((topLeft === bottomRight)) {
            throw new exception.ValueError("Rectangle has no size");
        }
        super();
        this.width = Math.abs(topLeft.x - bottomRight.x);
        this.height = Math.abs(topLeft.y - bottomRight.y);
        this.centre = new Point((topLeft.x + bottomRight.x)/2.0, (topLeft.y + bottomRight.y)/2.0);
        this.topLeft = this.centre.subtract([this.width/2.0, this.height/2.0]);
        this.bottomRight = this.centre.add([this.width/2.0, this.height/2.0]);
    }

    contains(point)
    {
        const offset = point.offset(this.centre);
        return 0 < offset[0] && offset[0] < this.width
            && 0 < offset[1] && offset[1] < this.height;
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'rect');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 x: this.topLeft.x, y: this.topLeft.y,
                                 width: this.width, height: this.height,
                                 fill: 'none'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class RoundedRectangle extends Rectangle
{
    constructor(topLeft, bottomRight, xCornerRadius = 0, yCornerRadius = 0)
    {
        super(topLeft, bottomRight);
        if (yCornerRadius === 0) {
            yCornerRadius = xCornerRadius;
        }
        if ((xCornerRadius < 0) || (xCornerRadius > this.width/2.0)
         || (yCornerRadius < 0) || (yCornerRadius > this.height/2.0)) {
            throw new exception.ValueError("Invalid corner radius");
        }
        this.xCornerRadius = xCornerRadius;
        this.yCornerRadius = yCornerRadius;
        if (xCornerRadius === 0 && yCornerRadius === 0) {
            this.innerRectangle = new Rectangle(topLeft, bottomRight);
            this.cornerEllipse = null;
        } else {
            const w_2 = (this.width - xCornerRadius)/2.0;
            const h_2 = (this.height - yCornerRadius)/2.0;
            this.innerRectangle = new Rectangle(new Point(this.centre.x).subtract([w_2, h_2]),
                                                new Point(this.centre.x).add([w_2, h_2]));
            this.cornerEllipse = new Ellipse(this.centre, xCornerRadius, yCornerRadius);
        }
    }

    contains(point)
    {
        if (xCornerRadius === 0 && yCornerRadius === 0) {
            return super.contains(point);
        } else {
            return this.innerRectangle.contains(point)
                || this.cornerEllipse.translate([-this.width/2.0, -this.height/2.0]).contains(point)
                || this.cornerEllipse.translate([-this.width/2.0,  this.height/2.0]).contains(point)
                || this.cornerEllipse.translate([ this.width/2.0, -this.height/2.0]).contains(point)
                || this.cornerEllipse.translate([ this.width/2.0,  this.height/2.0]).contains(point);
                // Or the side lobes...
        }
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'rect');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 x: this.topLeft.x, y: this.topLeft.y,
                                 width: this.width, height: this.height,
                                 rx: this.xCornerRadius, ry: this.yCornerRadius,
                                 fill: 'none'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class Ellipse extends GeoObject
{
    constructor(centre, xRadius, yRadius)
    {
        if ((xRadius === 0) || (yRadius === 0)) {
            throw new exception.ValueError("Invalid radius");
        }
        if (centre instanceof Array) {
            centre = new Point(...centre);
        }
        super();
        this.centre = centre;
        this.xRadius = xRadius;
        this.yRadius = yRadius;
    }

    line_intersect(line)
    {
        const l = (this.centre.x !== 0) || (this.centre.y !== 0)
                ? line.translate([-this.centre.x, -this.centre.y])
                : line;
        const points = [];
        if (l.A === 0) {
            let y = -l.C/l.B;
            const x2 = Math.pow(this.xRadius, 2) - Math.pow(y*this.xRadius/this.yRadius, 2);
            if (x2 === 0) {
                points.push(new Point(0, y).add([this.centre.x, this.centre.y]));
            } else if ( x2 > 0) {
                const x = Math.sqrt(x2);
                points.push(new Point(-x, y).add([this.centre.x, this.centre.y]));
                points.push(new Point( x, y).add([this.centre.x, this.centre.y]));
            }
        } else {
            const d2 = Math.pow(l.A*this.xRadius, 2) + Math.pow(l.B*this.yRadius, 2);
            const a = -l.C*l.B*Math.pow(this.yRadius, 2)/d2;
            const b = d2 - Math.pow(l.C, 2);
            if (b === 0) {
                points.push(new Point(-(l.C + a*l.B)/l.A, a).add([this.centre.x, this.centre.y]));
            } else if (b > 0) {
                const c = -l.A*this.xRadius*this.yRadius*Math.sqrt(b)/d2;
                points.push(new Point(-(l.C + (a + c)*l.B)/l.A, a + c).add([this.centre.x, this.centre.y]));
                points.push(new Point(-(l.C + (a - c)*l.B)/l.A, a - c).add([this.centre.x, this.centre.y]));
            }
        }
        return points.filter(p => line.contains(p));
    }

    contains(point)
    {
        return (Math.pow((point.x - this.centre.x)/this.xRadius, 2)
              + Math.pow((point.y - this.centre.y)/this.yRadius, 2) < 1.0);
    }

    translate(offset)
    {
        return new Ellipse(this.centre.add(offset), this.xRadius, this.yRadius);
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'ellipse');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 cx: this.centre.x, cy: this.centre.y,
                                 rx: this.xRadius, ry: this.yRadius,
                                 fill: 'none', stroke: 'blue'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export class Circle extends Ellipse
{
    constructor(centre, radius)
    {
        if (centre instanceof Array) {
            centre = new Point(...centre);
        }
        super(centre, radius, radius);
        this.radius = radius;
    }

    drawSvg(parentNode)
    {
        const svgNode = document.createElementNS(SVG_NS, 'circle');
        setAttributes(svgNode, { class: this.constructor.name.toLowerCase(),
                                 cx: this.centre.x, cy: this.centre.y,
                                 r: this.radius,
                                 fill: 'none', stroke: 'blue'});
        parentNode.appendChild(svgNode);
    }
}

//==============================================================================

export function test()
{
    const circles = [new Ellipse(new Point(200, 200), 150, 100),
                     new Circle(new Point(400, 200), 80),
                     new Circle(new Point(200, 400), 80)
                    ];

    const lines = [new LineSegment(new Point(200, 100), new Point(200, 300)),
                   new LineSegment(new Point(400, 100), new Point(400, 300)),
                   new LineSegment(new Point(100, 400), new Point(300, 400)),
                   new LineSegment(new Point(  0,   0), new Point(200, 200))
                  ];

    const lxy = new LineSegment(new Point(200, 300), new Point(400, 100));
    const pxy = new ProjectiveLine(lxy.A, lxy.B, lxy.C);
    lines.push(pxy);
    lines.push(pxy.translate([50, 50]));

    const elements = new List();
    elements.extend(circles).extend(lines);

    for (let l of lines) {
        for (let c of circles) {
            let s = [];
            for (let p of c.line_intersect(l)) {
                elements.push(p);
                s.push(p.toString());
            }
            console.log(s);
        }
    }

    return elements;
}

//==============================================================================
